### PR TITLE
Reject setting invalid versions on CSRs

### PR DESCRIPTION
**Pull request by job**
Reject setting invalid versions on CSRs

Discussed in https://github.com/openssl/openssl/issues/20663

### CODE REVIEW

**Review by Sashan — COMMENTED** [review]
The check does not hurt for sure. Though I still have a question:
currently the function is being called from `make_REQ()` only. Right here in `apps/req.c`:
```
1080
1081     /* so far there is only version 1 */
1082     if (!X509_REQ_set_version(req, X509_REQ_VERSION_1))
1083         goto err;
1084
1085     if (fsubj != NULL)
```
Are we adding extra safety belt here w.r.t. future changes? or do I miss some code path?

thanks for clarification.

**Comment by davidben** [comment]
The function is public API, with a bit of a history of folks being confused and generating invalid CSRs as a result.

**Review by Sashan — APPROVED** [review]
sorry I've missed the link to discussion. change looks good.
thanks @davidben for clarification.

File: crypto/x509/x509rset.c [hunk 1]
```diff
@@ -19,6 +19,10 @@ int X509_REQ_set_version(X509_REQ *x, long version)
 {
     if (x == NULL)
         return 0;
+    if (version != X509_REQ_VERSION_1) {
+        ERR_raise(ERR_LIB_X509, ERR_R_PASSED_INVALID_ARGUMENT);
```
Comment (t8m) [inline crypto/x509/x509rset.c:23]: Could you please merge this with the previous `if` as I think raising the same error for `x == NULL` would be also beneficial.

**Review by t8m — CHANGES_REQUESTED** [review]

**Review by t8m — APPROVED** [review]

**Review by t-j-h — APPROVED** [review]

**Comment by nhorman** [comment]
merged, thank you for your contribution

**Comment by ibgp2** [comment]
Currently, Debian testing ships `openssl-3.5.4-1` and `intune-portal` doesn't work:

> Couldn't enroll your device: X509_REQ_set_version:passed invalid argument:crypto/x509/x509set.c

According to [this link](https://github.com/recolic/microsoft-intune-archlinux/blob/master/README.md) this bug is introduced and impact `openssl>3.3.4`.

> **Couldn't enroll your device: X509_REQ_set_version:passed invalid argument:crypto/x509/x509set.c**
> 
> Same as previous issue.
> Downgrade your openssl to 3.3.4 or older. Please use fix-libssl.sh instead of your package manager to avoid breaking other programs.
> [Error Screenshot](https://github.com/recolic/microsoft-intune-archlinux/blob/master/.res/2.png)

The previous mentioned issue claims the bug was introduced by [this PR](https://github.com/openssl/openssl/pull/23965):
 
> https://github.com/openssl/openssl/pull/23965
> 
> at this commit (included since openssl 3.4.0):
> 397051a40db2d68433b842e7505e8cf3c9effb36 (main)
> 
> Observed regression in other projects such as:
> https://github.com/ruby/openssl/issues/734
> 
> Solution 1: Downgrade to openssl-3.3.4
> Solution 2: Write libssl_fix.so with a good version of that tiny function, use LD_PRELOAD to shadow the original buggy impl.

This bug looks [confirmed also here](https://console.systems/2024/03/how-to-install-intune-in-debian-12.html).
I'm not an expert and don't know if it is true, but if so, could you fix it?

Best regards

**Comment by davidben** [comment]
That sounds like a bug in intune-portal. There is only one CSR version defined, so `X509_REQ_set_version` should only ever be passed one value. That value is actually the default, so they actually should just delete the function call.

**Pull request closed without merging**